### ISSUE TITLE
CoverageData with no_disk does not update correctly

### ISSUE THREAD

**Issue by lordmauve** [coveragepy/coveragepy#1323]
**Describe the bug**

A CoverageData constructed with `no_disk=True` cannot be updated with another CoverageData. The operation succeeds but records nothing:

```python
>>> from coverage import CoverageData
>>> a = CoverageData('.foo')         
>>> a.add_arcs({'foo.py': {}})       
>>> a.measured_files()               
{'foo.py'}                           
                                     
>>> b = CoverageData(no_disk=True)   
>>> b.update(a)                      
>>> b.measured_files()               
set()                                
```

(This applies also for arcs etc., `b` remains empty).

**Expected behavior**

The expected behaviour for `no_disk=True` is the same as when `no_disk` is left unspecified; after an `update()` call we should see data in the updated CoverageData:

```python
>>> c = CoverageData()
>>> c.update(a)       
>>> c.measured_files()
{'foo.py'}
```

**To Reproduce**

1. What version of Python are you using? *3.9.7*
1. What version of coverage.py shows the problem? *6.3.1*
1. The output of `coverage debug sys` is helpful.

It's very long due to a tremendously long `PYTHONPATH`, but perhaps this part is most relevant:

```
                sqlite3_version: 2.6.0                                                 
         sqlite3_sqlite_version: 3.36.0                                                
             sqlite3_temp_store: 0                                                     
        sqlite3_compile_options: COMPILER=gcc-8.3.0; ENABLE_FTS4; ENABLE_FTS5          
                                 ENABLE_GEOPOLY; ENABLE_JSON1; ENABLE_MATH_FUNCTIONS   
                                 ENABLE_RTREE; MAX_MMAP_SIZE=17179869184; SYSTEM_MALLOC
                                 THREADSAFE=1                                          
```

**Comment by nedbat** [comment]
Thanks for the interesting bug report.  Just for my own curiosity, can you say more about how you are using these APIs?

**Comment by lordmauve** [comment]
We do a huge test run of our monorepo with tests parallelised across a test farm. The coverage files get rsynced back to one host where we:

1. Merge them
2. Remap paths captured in the coverage files to paths that exists on that host at that time
3. Build HTML reports
4. Build a table of the coverage statistics per component

I think if you're asking whether this could just use the CLI, I think the command line doesn't let me supply the regex-based path mapper class we use for step 2.

**Comment by lordmauve** [comment]
Oh, the reason I was reluctant to use a disk file was that a) we don't keep that file, no need for it to be on disk and b) for some reason this is happening on a read-only filesystem

**Comment by nedbat** [comment]
Hmm, I can confirm this happens, but it's a mystery (so far) to me why...

**Comment by nedbat** [comment]
This is now fixed in 84f70f69.

**Comment by nedbat** [comment]
This is now released as part of [coverage 6.4](https://pypi.org/project/coverage/6.4).
